packages/flutter/lib/src/material/material_state.dart

```diff
@@ -674,3 +674,44 @@ class _MaterialStatePropertyAll<T> implements MaterialStateProperty<T> {
   @override
   String toString() => 'MaterialStateProperty.all($value)';
 }
+
+
+/// Manages a set of [MaterialState]s and notifies listeners of changes.
+///
+/// Used by widgets that expose their internal state for the sake of
+/// extensions that add support for additional states. See
+/// [TextButton.statesController] for example.
+///
+/// The controller's [value] is its current set of states. Listeners
+/// are notified whenever the [value] changes. The [value] should only be
+/// changed with [add] and [remove], it should not be modified directly.
+class MaterialStatesController extends ValueNotifier<Set<MaterialState>> {
+  /// Creates a MaterialStateController.
+  MaterialStatesController([Set<MaterialState>? value]) : super(<MaterialState>{...?value});
+
+  /// Adds [state] to [value] and notifies listeners if the [value]
+  /// has changed.
+  void add(MaterialState state) {
+    if (!value.contains(state)) {
+      value = <MaterialState>{state, ...value};
+    }
+  }
+
+  /// Removes [state] from [value] and notifies listeners if the [value]
+  /// has changed.
+  void remove(MaterialState state) {
+    if (value.contains(state)) {
+      value = value.difference(<MaterialState>{state});
+    }
+  }
+
+  /// Applies [add] to [state] if [addState] is true,
+  /// [remove] otherwise.
+  void update(MaterialState state, bool addState) {
+    if (addState) {
+      add(state);
+    } else {
+      remove(state);
+    }
+  }
```

Comment:
Just supporting MaterialStatesController.update() is an excellent suggestion; I've done that. Internally, using expressions like `value = value..add(state);` to update `value` is problematic because the initial value might be unmodifiable.

Reply:
Gotcha, makes sense! Thanks for checking all these suggestions out.

And by the way, I'm realizing that I might not have permissions to actually resolve all of these comment threads, so if having them open bothers you at all, feel free to do so 👍🏻

Reply:
Thanks for all of your comments BTW, the API and the implementation are much improved for them.